source/util/move_to_front.cpp

```diff
@@ -0,0 +1,336 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "move_to_front.h"
+
+#include <algorithm>
+#include <iomanip>
+#include <iostream>
+#include <sstream>
+
+namespace spvutils {
+
+size_t MoveToFront::RankFromId(uint32_t id) {
+  // The id was previously deprecated (i.e left its scope).
+  assert(deprecated_ids_.count(id) == 0);
+
+  const size_t old_size = GetSize();
+  const auto it = id_to_node_.find(id);
+
+  if (it == id_to_node_.end()) {
+    InsertNode(CreateNode(next_timestamp_++, id));
+    assert(old_size + 1 == GetSize());
+    return old_size;
+  }
+
+  uint16_t target = it->second;
+
+  uint16_t node = target;
+  size_t rank = SizeOf(LeftOf(node));
```

Comment:
I don't understand.  uint16_t is a valid type, implemented by the compiler.

And the code remains unchanged, i.e. no static_cast<uint16>(...).    In any case, this is a minor point.